profiler/src/Demos/Samples.Computer01/SocketTimeout.cs

```diff
@@ -0,0 +1,175 @@
+                _serverReadyEvent.Wait();
+            }
+
+            if (IsEventSet())
+            {
+                return;
+            }
+
+            ReceiveTimeout();
+            SendTimeout();
+        }
+
+        private static Socket CreateServerSocket()
+        {
+            var s = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+            var ep = new IPEndPoint(IPAddress.Any, ServerPort);
+            s.Bind(ep);
+            s.Listen();
+
+            return s;
+        }
+
+        private Socket CreateAndConnectClientSocket()
+        {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp)
+            {
+                SendTimeout = OperationTimeout,
+                ReceiveTimeout = OperationTimeout,
+                SendBufferSize = 0,
+                Blocking = true
+            };
+            socket.Connect(_serverEndpoint);
+
+            return socket;
+        }
+
+        private void ReceiveTimeout()
+        {
+            EnsureOperationTimesOut(
+                (socket) =>
+                {
+                    var buffer = new byte[42];
+                    socket.Receive(buffer);
+                },
+                "Receive");
+        }
+
+        private void SendTimeout()
+        {
+            EnsureOperationTimesOut(
+                (socket) =>
+                {
+                    // really big buffer to make sure the call blocks
+                    var buffer = new byte[200_000];
+                    socket.Send(buffer);
+                },
+                "Send");
+        }
+
+        private bool IsSocketTimeoutException(Exception e)
+        {
+            return e switch
+            {
+                AggregateException ae => IsSocketTimeoutException(ae.InnerException),
+                SocketException se => se.ErrorCode == TimeoutErrorCode,
+                _ => false
+            };
+        }
+
+        private void EnsureOperationTimesOut(Action<Socket> action, string operationName)
+        {
+            try
+            {
+                var cts = new CancellationTokenSource(_operationErrorTimeout);
+                var task = Task.Run(
+                    () =>
+                    {
+                        using var socket = CreateAndConnectClientSocket();
+                        action(socket);
+
+                        if (!IsEventSet())
+                        {
+                            Console.WriteLine($"[Error] The {operationName} blocking call must timeout.");
+                        }
+                    });
+                task.Wait(cts.Token);
+            }
+            catch (Exception e) when (IsSocketTimeoutException(e))
+            {
+                // ok case
+            }
+            catch (OperationCanceledException)
+            {
+                Console.WriteLine($"[Error] The {operationName} operation was cancelled because it reached the 2s timeout.");
```

Comment:
small detail but _operationErrorTimeout is a 3 seconds timeout; not 2 seconds

Reply:
👍